### PR TITLE
MTD: store geometry scenario in topology record

### PR DESCRIPTION
**Pull request by fabiocos**
Several MTD geometry scenarios are under study for the TDR, and they differ not only in the pure geometrical description of volumes (encoded in DD) but also in some topological parameters (e.g. the number of modules per type and total number of modules per rod). This information is needed in the event to pick the appropriate numbers (hardcoded in BTLDetId for the time being to support eta and phi methods) for geometry computation in digitization and reconstruction.

This class adds this information in mtdStructureTopology.xml files, which could be used to store the whole topology information (as provided by tkLayout for tracker). The mechanism used to propagate the information though the MTDParametersFromDD class mimics what is done in the Hcal geometry though the HcalTopologyMode object (from which the code has been reused).

A simple test class it provided that shows how the correct scenario is picked and can be retrieved through a MTDTopology accessor method. This test class can be further extended to validate other functionalities (beyond the scop of this PR).

### CODE REVIEW

**Comment by fabiocos** [comment]
@pmeridian @abenagli FYI
@casarsa you should use this to finalize your PR about bar digitization

**Comment by fabiocos** [comment]
please test workflow 27434.0

**Comment by fabiocos** [comment]
Running https://github.com/cms-sw/cmssw/pull/25352/files#diff-7690999b4b23daf922a2fa1778b04bfc with geometry scenario D35 provides as output:

Begin processing the 1st record. Run 1, Event 1, LumiSection 1 on stream 0 at 27-Nov-2018 15:58:08.071 CET
TimeModule> 1 1 prod GeometricTimingDetAnalyzer 11.7403
MTDTopologyAnalyzer: Here I am
MTDTopologyAnalyzer: MTD topology mode = 3
TimeModule> 1 1 prod1 MTDTopologyAnalyzer 0.00730395

while out of the box (scenario D24) it gives

Begin processing the 1st record. Run 1, Event 1, LumiSection 1 on stream 0 at 27-Nov-2018 16:00:05.676 CET
TimeModule> 1 1 prod GeometricTimingDetAnalyzer 18.1132
MTDTopologyAnalyzer: Here I am
MTDTopologyAnalyzer: MTD topology mode = 1
TimeModule> 1 1 prod1 MTDTopologyAnalyzer 0.00771403

as expected

File: Geometry/MTDNumberingBuilder/test/MTDTopologyAnalyzer.cc [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+
+#include "FWCore/ParameterSet/interface/ParameterSet.h"
+#include "Geometry/Records/interface/MTDTopologyRcd.h"
+#include "Geometry/MTDNumberingBuilder/interface/MTDTopology.h"
+
+#include "FWCore/MessageLogger/interface/MessageLogger.h"
+
+//
+//
+// class decleration
```
Comment (vargasa) [inline Geometry/MTDNumberingBuilder/test/MTDTopologyAnalyzer.cc]: Do we really need this comment? *declaration*

File: Geometry/MTDNumberingBuilder/test/MTDTopologyAnalyzer.cc [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+{
+public:
+      explicit MTDTopologyAnalyzer( const edm::ParameterSet& );
+      ~MTDTopologyAnalyzer() override;
+
+  void beginJob() override {}
+  void analyze(edm::Event const& iEvent, edm::EventSetup const&) override;
+  void endJob() override {}
+
+private:
```
Comment (vargasa) [inline Geometry/MTDNumberingBuilder/test/MTDTopologyAnalyzer.cc]: no `private` members being declared. Remove

File: Geometry/MTDNumberingBuilder/test/MTDTopologyAnalyzer.cc [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+};
+
+MTDTopologyAnalyzer::MTDTopologyAnalyzer( const edm::ParameterSet& iConfig )
+{}
+
+MTDTopologyAnalyzer::~MTDTopologyAnalyzer()
+{}
```
Comment (vargasa) [inline Geometry/MTDNumberingBuilder/test/MTDTopologyAnalyzer.cc]: Use `= default` instead

File: Geometry/MTDNumberingBuilder/test/MTDTopologyAnalyzer.cc [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+
+// ------------ method called to produce the data  ------------
+void
+MTDTopologyAnalyzer::analyze( const edm::Event& iEvent, const edm::EventSetup& iSetup )
+{
+
+  PRINT("MTDTopologyAnalyzer")<< "Here I am" << std::endl;
```
Comment (vargasa) [inline Geometry/MTDNumberingBuilder/test/MTDTopologyAnalyzer.cc]: I think you need to remove this `Here I am` ;)

File: Geometry/MTDNumberingBuilder/test/MTDTopologyAnalyzer.cc [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+
+ Description: <one line class summary>
+
+ Implementation:
+     <Notes on implementation>
+*/
+//
+
+// system include files
+#include <memory>
```
Comment (vargasa) [inline Geometry/MTDNumberingBuilder/test/MTDTopologyAnalyzer.cc]: Are you sure you are using all of these `#include`s ? Some of them are not being used in this file

File: Geometry/MTDCommonData/data/CrystalBar/mtdStructureTopology.xml [hunk 6]
```diff
@@ -0,0 +1,13 @@
+<?xml version="1.0"?>
+<DDDefinition xmlns="http://www.cern.ch/cms/DDL" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://www.cern.ch/cms/DDL ../../../DetectorDescription/Schema/DDLSchema.xsd">
```
Comment (vargasa) [inline Geometry/MTDCommonData/data/CrystalBar/mtdStructureTopology.xml]: I'm missing something here. Is this a relative path?

Comment (fabiocos) [reply Geometry/MTDCommonData/data/CrystalBar/mtdStructureTopology.xml]: @vargasa this is the usual xml header of all the DD geometry files, referring to https://cmssdt.cern.ch/lxr/source/DetectorDescription/Schema/DDLSchema.xsd
For its structure @ianna can probably better comment, I guess it is historical

**Comment by fabiocos** [comment]
@vargasa your comments have been addressed in the latest commit, cleaning the class (which was just copied and adjusted from another example)

**Comment by fabiocos** [comment]
the latest change has no effect on the PR test

@smuzaffar the header test failure looks unrelated to the changes in this PR, and looks spurious to me given the type of error message, do you agree?

**Comment by smuzaffar** [comment]
@fabiocos , yes please ignore these header checks errors. I still need to understand why we see these ICE for header checks.

**Comment by fabiocos** [comment]
@ianna @kpedro88 @ggovi comments?

**Comment by kpedro88** [comment]
please test

**Comment by kpedro88** [comment]
+upgrade

**Comment by cvuosalo** [comment]
+1

**Comment by fabiocos** [comment]
@ggovi @lpernie @tocheng do you have comments? Otherwise I would like to merge this PR, as I am working on its second part to replace mtdParameters in PSet with xml (as done for other detectors)

**Comment by ggovi** [comment]
There is a schema change in the CondFormats PMTDParameters. However, no payload have been stored so far with instances of this class.  So, the change can be allowed but bear in mind that the previous PRE releases might not be able to read the future payloads.

**Comment by ggovi** [comment]
+1

**Comment by lpernie** [comment]
+1

**Comment by fabiocos** [comment]
@ggovi thanks for the comment. In 10_4_0_pre2 GEN-SIM will be produced, but this CondFormat is not really used there, and in any case no payload is made so far for this information, as it is still at the level of R&D test

**Comment by fabiocos** [comment]
+1

**Pull request merged**